pkg/referrerstore/oras/authprovider/k8secret_authprovider.go

```diff
@@ -82,122 +84,139 @@ func (s *k8SecretProviderFactory) Create(authProviderConfig AuthProviderConfig)
 	}
 
 	// get name of namespace ratify is running in
-	ratifyNamespace := os.Getenv(ratifyNamespaceEnv)
-	if ratifyNamespace == "" {
+	namespace := os.Getenv(ratifyNamespaceEnv)
+	if namespace == "" {
 		return nil, fmt.Errorf("environment variable %s not set", ratifyNamespaceEnv)
 	}
 
-	var k8secrets []*core.Secret
-	ctx := context.Background()
+	return &k8SecretAuthProvider{
+		ratifyNamespace:  namespace,
+		config:           conf,
+		clusterClientSet: clientSet,
+	}, nil
+}
+
+// Enabled checks if ratify namespace, config, or cluster client set is nil
+func (d *k8SecretAuthProvider) Enabled(ctx context.Context) bool {
+	if d.ratifyNamespace == "" || d.clusterClientSet == nil {
+		return false
+	}
+
+	return true
+}
+
+// Provide finds secret corresponding to artifact's registry host name, extracts
+// the authentication credentials from k8 secret, and returns AuthConfig
+func (d *k8SecretAuthProvider) Provide(ctx context.Context, artifact string) (AuthConfig, error) {
+	if !d.Enabled(ctx) {
+		return AuthConfig{}, fmt.Errorf("k8 auth provider not properly enabled")
+	}
+
+	hostName, err := GetRegistryHostName(artifact)
+	if err != nil {
+		return AuthConfig{}, err
+	}
 
 	// iterate through config secrets, resolve each secret, and store in map
-	for _, k8secret := range conf.Secrets {
+	for _, k8secret := range d.config.Secrets {
 		// default value of secret is assumed to be ratify namespace
 		if k8secret.Namespace == "" {
-			k8secret.Namespace = ratifyNamespace
+			k8secret.Namespace = d.ratifyNamespace
 		}
 
-		secret, err := clusterClientSet.CoreV1().Secrets(k8secret.Namespace).Get(ctx, k8secret.SecretName, meta.GetOptions{})
+		secret, err := d.clusterClientSet.CoreV1().Secrets(k8secret.Namespace).Get(ctx, k8secret.SecretName, meta.GetOptions{})
 		if err != nil {
-			return nil, err
+			return AuthConfig{}, err
 		}
 
 		// only dockercfg or docker config json secret type allowed
 		if secret.Type == core.SecretTypeDockercfg || secret.Type == core.SecretTypeDockerConfigJson {
-			k8secrets = append(k8secrets, secret)
+			authConfig, err := d.resolveCredentialFromSecret(hostName, secret)
+			if err != nil {
+				return AuthConfig{}, err
+			}
+			// if a non empty AuthConfig is returned
+			if authConfig != (AuthConfig{}) {
```

Comment:
What are the cases where auth config is empty? wouldn't it be an error case that will be caught in the line 135?

Reply:
`resolveCredentialFromSecret` would return an empty `AuthConfig` without an error when it can properly resolve the k8 secret but it can't find a corresponding registry hostname credential entry in the secret.

Reply:
Yes I'll switch it to throw an error. And then it can be caught and dealt with accordingly in Provide